flutter-engine/src/texture_registry.rs

```diff
@@ -1,163 +1,88 @@
-    name: GLuint,
-    width: u32,
-    height: u32,
+    textures: Arc<Mutex<HashMap<TextureId, ExternalTexture>>>,
+    data: Arc<Mutex<HashMap<TextureId, RgbaImage>>>,
 }
 
 impl TextureRegistry {
     pub fn new() -> Self {
         Default::default()
     }
 
-    pub fn create_texture(&mut self, engine: &FlutterEngine) -> Arc<ExternalTexture> {
+    pub fn create_texture(&self, engine: &FlutterEngine, img: RgbaImage) -> TextureId {
         static TEXTURE_ID: AtomicI64 = AtomicI64::new(1);
-
         let texture_id = TEXTURE_ID.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
-        let flutter_texture = engine.register_external_texture(texture_id);
 
-        let texture = Arc::new(ExternalTexture {
-            texture: flutter_texture,
-            texture_data: Mutex::new(None),
-        });
+        self.data.lock().unwrap().insert(texture_id, img);
 
-        self.textures
-            .initial
-            .insert(texture_id, Arc::clone(&texture));
-
-        texture
+        let textures = self.textures.clone();
+        engine.run_on_platform_thread(move |engine| {
+            let texture = engine.register_external_texture(texture_id);
+            textures.lock().unwrap().insert(texture_id, texture);
+        });
+        texture_id
     }
 
     pub fn get_texture_frame(
-        &mut self,
-        texture_id: TextureID,
-        (width, height): (u32, u32),
+        &self,
+        texture_id: TextureId,
+        _size: (usize, usize),
     ) -> Option<ExternalTextureFrame> {
-        if let Some(texture) = self.textures.initial.remove(&texture_id) {
-            // texture is still initial --> create it
-            let frame = unsafe { create_gl_texture(texture_id, (width, height)) };
-            let mut data = texture.texture_data.lock().unwrap();
-            data.replace(TextureData {
-                name: frame.name,
-                width,
-                height,
-            });
-            self.textures
-                .created
-                .insert(texture_id, Arc::downgrade(&texture));
-            return Some(frame);
-        }
-
-        if let Some(texture) = self.textures.created.get(&texture_id) {
-            // texture has been created, this is a notification that a new frame is available
-            if let Some(_texture) = texture.upgrade() {
-                // texture still alive
-                // TODO: notify of new frame
-            } else {
-                // texture was dropped, remove it from here as well
-                self.textures.created.remove(&texture_id);
-            }
-        }
-        None
-    }
-}
-
-impl ExternalTexture {
-    pub fn handle(&self) -> TextureID {
-        self.texture.texture_id
-    }
-
-    pub fn gl_texture(&self) -> Option<GLuint> {
-        let data = self.texture_data.lock().unwrap();
-        data.as_ref().map(|data| data.name)
+        let data = self.data.lock().unwrap();
+        let img = data.get(&texture_id).unwrap();
```

Comment:
We should probably return `None` if the texture isn't found, so a simple bug like using `Texture()`  with the wrong id doesn't panic and bring down the entire app